### PR TITLE
a generic pump plugin

### PR DESCRIPTION
**Pull request by jasoncalabrese**
screen shot 2016-01-05 at 6 51 52 pm

Includes battery voltage and reservoir level based alarms

### CODE REVIEW

**Comment by jasoncalabrese** [comment]
@MilosKozak think you also have a pump plugin in progress, we should get them merged together soon.

File: lib/plugins/pump.js [hunk 1]
```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+var _ = require('lodash');
+var moment = require('moment');
+var times = require('../times');
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ['times' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1721301845/issues/source?bid=2452379&fileBranchId=3051223#l5)

File: lib/plugins/pump.js [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+var levels = require('../levels');
+var timeago = require('./timeago')();
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ['timeago' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1721301845/issues/source?bid=2452379&fileBranchId=3051223#l7)

File: lib/plugins/pump.js [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+      , label: 'Res'
+      , info: info
+    });
+  };
+
+  return pump;
+
+}
+
+function checkStatus (prop, prefs) {
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1721301845/issues/source?bid=2452379&fileBranchId=3051223#l81)

File: lib/plugins/pump.js [hunk 4]
```diff
@@ -0,0 +1,121 @@
+'use strict';
+
+var _ = require('lodash');
+var moment = require('moment');
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ['moment' is defined but never used.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1721301845/issues/source?bid=2452379&fileBranchId=3051223#l4)

**Comment by MilosKozak** [comment]
I'd prefer to see timeago as a main value. Connection to DanaR is closed after comm to save battery. The rest of values in tooltip
This is my approach
image
image

**Comment by jasoncalabrese** [comment]
I was thinking it would be nice it we can pick the field(s) we display in the pill, would be easy as an env var, but really would be good to set it at from the client.

**Comment by MilosKozak** [comment]
Do you mean the main value?

**Comment by MilosKozak** [comment]
Values in tooltip can be added when appear in status record

**Comment by MilosKozak** [comment]
are you able to convert voltage to battery icon?

**Comment by jasoncalabrese** [comment]
Yeah, I think the main value in the pill should be configurable, so we can make the plugin fit several uses.  The battery voltage is a little crazy on the MM pump, it jump around a lot up and down when using lithium batteries.  They stay strong till the end then drop really fast.  Even the display on the pump goes up and down. Range for the battery seems to be 1.6x to 1.3x

File: lib/plugins/pump.js (hunk 3 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743842086/issues/source?bid=2452379&fileBranchId=3051223#l81)

File: lib/plugins/pump.js (hunk 4 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["moment" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743842086/issues/source?bid=2452379&fileBranchId=3051223#l4)

File: lib/plugins/pump.js (hunk 2 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["timeago" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743842086/issues/source?bid=2452379&fileBranchId=3051223#l7)

File: lib/plugins/pump.js (hunk 1 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["times" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743842086/issues/source?bid=2452379&fileBranchId=3051223#l5)

File: lib/plugins/pump.js [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+  if (level === levels.WARN) {
+    cls = 'warn';
+  } else if (level === levels.URGENT) {
+    cls = 'urgent';
+  }
+
+  return cls;
+}
+
+function checkStatus (prop, prefs) {
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743892576/issues/source?bid=2452379&fileBranchId=3051223#l99)

File: lib/plugins/pump.js [hunk 6]
```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+var _ = require('lodash');
+var moment = require('moment');
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["moment" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743892576/issues/source?bid=2452379&fileBranchId=3051223#l4)

File: lib/plugins/pump.js [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+var times = require('../times');
+var levels = require('../levels');
+var timeago = require('./timeago')();
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["timeago" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743892576/issues/source?bid=2452379&fileBranchId=3051223#l7)

File: lib/plugins/pump.js [hunk 8]
```diff
@@ -0,0 +1,141 @@
+'use strict';
+
+var _ = require('lodash');
+var moment = require('moment');
+var times = require('../times');
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["times" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1743892576/issues/source?bid=2452379&fileBranchId=3051223#l5)

File: lib/client/renderer.js [hunk 9]
```diff
@@ -191,7 +194,9 @@ function init (client, d3) {
       var notes = treatment.notes || '';
       var enteredBy = treatment.enteredBy || '';
 
-      var notOpenAPSSpam = enteredBy.indexOf('openaps://') === -1 || (notes.indexOf('BasalProfileStart') === -1 && notes.indexOf('ResultDailyTotal') === -1);
+      var notOpenAPSSpam = enteredBy.indexOf('openaps://') === -1 || _.isUndefined(_.find(openAPSSpam, function startsWith (spam) {
+        return notes.indexOf(spam) === 0
```
Comment (jasoncalabrese) [inline lib/client/renderer.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1804879484/issues/source?bid=2452379&fileBranchId=3051223#l198)

File: lib/plugins/pump.js (hunk 6 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["moment" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1804879521/issues/source?bid=2452379&fileBranchId=3051223#l4)

File: lib/plugins/pump.js (hunk 7 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["timeago" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1804879521/issues/source?bid=2452379&fileBranchId=3051223#l7)

File: lib/plugins/pump.js (hunk 8 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["times" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1804879521/issues/source?bid=2452379&fileBranchId=3051223#l5)

File: lib/plugins/pump.js (hunk 5 above)
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (8)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1804879521/issues/source?bid=2452379&fileBranchId=3051223#l99)

File: lib/plugins/pump.js [hunk 10]
```diff
@@ -0,0 +1,266 @@
...
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: ["times" is defined but never used](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1808302552/issues/source?bid=2452379&fileBranchId=3051223#l5)

File: lib/plugins/pump.js [hunk 11]
```diff
@@ -0,0 +1,266 @@
...
+  var pump = {
+    name: 'pump'
+    , label: 'Pump'
+    , pluginType: 'pill-status'
+  };
+
+  pump.getPrefs = function getPrefs (sbx) {
+
+    function cleanList (value) {
+      return decodeURIComponent(value || '').toLowerCase().split(' ')
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [Missing semicolon.](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1808302552/issues/source?bid=2452379&fileBranchId=3051223#l22)

File: lib/plugins/pump.js [hunk 12]
```diff
@@ -0,0 +1,265 @@
...
+
+function init ( ) {
+
+  var pump = {
+    name: 'pump'
+    , label: 'Pump'
+    , pluginType: 'pill-status'
+  };
+
+  pump.getPrefs = function getPrefs (sbx) {
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (11)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1808382488/issues/source?bid=2452379&fileBranchId=3051223#l18)

File: lib/plugins/pump.js [hunk 13]
```diff
@@ -0,0 +1,265 @@
...
+
+    if (result.reservoir) {
+      message.push('Pump Reservoir: ' + result.reservoir.display);
+    }
+
+    result.message = message.join('\n');
+  }
+}
+
+function prepareData (prop, prefs, sbx) {
```
Comment (jasoncalabrese) [inline lib/plugins/pump.js]: Codacy Issue found: [This function's cyclomatic complexity is too high. (10)](https://www.codacy.com/app/Nightscout/cgm-remote-monitor/file/1808382488/issues/source?bid=2452379&fileBranchId=3051223#l201)

**Comment by jasoncalabrese** [comment]
I think this is done and ready for extra testing.  I have it running on my prod site and will merge tomorrow unless we find an issue.  Take a look at the pump plugin in readme for all the options: https://github.com/nightscout/cgm-remote-monitor/tree/wip/pump#built-inexample-plugins

**Pull request merged**